langserver/workspace_refs.go

```diff
@@ -0,0 +1,173 @@
+			if err != nil {
+				log.Printf("externalRefsFromPkg: %v: %v\n", pkg, err)
+			}
+		}(pkg)
+	}
+	run.Wait()
+
+	sort.Sort(&results) // sort to provide consistent results
+
+	// TODO: We calculate all the results and then throw them away. If we ever
+	// decide to begin using limiting, we can improve the performance of this
+	// dramatically. For now, it lives in the spec just so that other
+	// implementations are aware it may need to be done and to design with that
+	// in mind.
+	if len(results.results) > params.Limit && params.Limit > 0 {
+		results.results = results.results[:params.Limit]
+	}
+	return results.results, nil
+}
+
+// externalRefsFromPkg collects all the external references from the specified
+// package and returns the results.
+func (h *LangHandler) externalRefsFromPkg(ctx context.Context, bctx *build.Context, conn JSONRPC2Conn, pkg string, rootPath string, results *refResultSorter) (err error) {
+	span, ctx := opentracing.StartSpanFromContext(ctx, "externalRefsFromPkg")
+	defer func() {
+		if err != nil {
+			ext.Error.Set(span, true)
+			span.SetTag("err", err.Error())
+		}
+		span.Finish()
+	}()
+	span.SetTag("pkg", pkg)
+
+	// Import the package.
+	buildPkg, err := bctx.Import(pkg, rootPath, 0)
+	if err != nil {
+		if !(strings.Contains(err.Error(), "no buildable Go source files") || strings.Contains(err.Error(), "found packages") || strings.HasPrefix(pkg, "github.com/golang/go/test/")) {
+			return fmt.Errorf("skipping possible package %s: %s\n", pkg, err)
+		}
+		return nil
+	}
+
+	pkgInWorkspace := func(path string) bool {
+		return PathHasPrefix(path, h.init.RootImportPath)
+	}
+
+	// Perform type checking.
+	fs, prog, diags, err := h.cachedTypecheck(ctx, bctx, buildPkg)
+	if err != nil {
+		return err
+	}
+	if len(diags) > 0 {
+		if err := h.publishDiagnostics(ctx, conn, diags); err != nil {
+			return fmt.Errorf("sending diagnostics: %s", err)
+		}
+	}
+
+	// Compute external references.
+	cfg := &refs.Config{
+		FileSet:  fs,
+		Pkg:      prog.Package(pkg).Pkg,
+		PkgFiles: prog.Package(pkg).Files,
+		Info:     &prog.Package(pkg).Info,
+	}
+	refsErr := cfg.Refs(func(r *refs.Ref) {
+		// If the reference itself is located in a vendor folder, exclude it.
+		locFilename := r.Position.Filename
+		if strings.HasPrefix(locFilename, "vendor/") || strings.Contains(locFilename, "/vendor/") {
+			return
+		}
+
+		var defName, defContainerName string
+		if fields := strings.Fields(r.Def.Path); len(fields) > 0 {
+			defName = fields[0]
+			defContainerName = strings.Join(fields[1:], " ")
+		}
+
+		defPkg, err := bctx.Import(r.Def.ImportPath, rootPath, build.FindOnly)
+		if err != nil {
+			// Log the error, and flag it as one in the trace -- but do not
+			// halt execution (hopefully, it is limited to a small subset of
+			// the data).
+			ext.Error.Set(span, true)
+			err := fmt.Errorf("externalRefsFromPkg: failed to import %v: %v", r.Def.ImportPath, err)
```

Comment:
If you run this on a large repo like kubernetes that has a lot of non-Go dirs, is this going to spit out a bunch of low-value error logs? I had to suppress a similar log call for that very reason, but not sure if those situations would be filtered out prior to reaching this log call here.